proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -163,9 +164,14 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
 		trace.Reset()
 
 		if reqInfo.RouteServiceURL == nil {
+			if attempt > 1 {
+				if attempt > reqInfo.RoutePool.NumEndpoints() && reqInfo.RoutePool.LastUpdated().Equal(lastUpdated) {
+					break
+				}
+			}
```

Comment:
Why do we need the call to `reqInfo.RoutePool.LastUpdated()` here? The first check should always return the current number of endpoints and since `lastUpdated` is never written within the loop it will always be false once the pool was updated, therefore causing an infinite loop.


I'd suggest:
```suggestion
			if attempt > 1 && attempt > reqInfo.RoutePool.NumEndpoints() {
				break
			}
```

Reply:
Yes as Amelia mentioned, just checking for number of instances is not enough. We want to account for a case when instance is registered and old instance is removed. This is when the issue is happening, after old instance is removed we fail to connect to it. And during rolling deploy you will have a replacement instance up, so the total number will be the same.

Reply:
Ah I see, from the title I was assuming this is about an app being scaled while gorouter is making requests.

My main point still stands though, assuming the time of the last updated time on the pool changes, `reqInfo.RoutePool.LastUpdated().Equal(lastUpdated)` will _always_ return `false` as we never update the `lastUpdated` which we compare against within the loop and it will always contain the initial value. So, unless I missed something, whenever that happens it would never stop looping.

Reply:
I'm trying to come up with a way to fix this but I'm not sure there is a straightforward way. To be on the safe side I wouldn't want to rely on the timestamp as that does't tell us how many instances changed and it could also be the case that there was an instance removed without one being added in which case we wouldn't want to try that endpoint knowing that its gone (if it's one we haven't tried yet).

So I guess without an additional variable we can't make this work. If we had an additional variable, we could use it to store the identity of the endpoints we already tried. When picking the next endpoint the iterator would check whether there is an endpoint we haven't tried and return that.

I've built a small demo but I'm not sure I like it: https://go.dev/play/p/7YuzIoCsbI5 which then outputs:

```
selected endpoint ep1
doesn't work, try again
selected endpoint ep2
doesn't work, try again
selected endpoint ep3
works, stopping
```

I guess what I want to say is that I rather see this as a feature of the iterator, not the round-tripper.

Reply:
`reqInfo.RoutePool.LastUpdated().Equal(lastUpdated)` will return true if the pool was updated during the loop run. Whenever we put something in the pool or remove an item we update this value - https://github.com/cloudfoundry/gorouter/blob/6c6bd520ad939910fe8c0830558f6dd179e7d113/route/pool.go#L298

These are all the test case scenarios that cover this (1. not updated, 2. updated with added new backends, 3. updated with 1 added and 1 removed backend) - https://github.com/cloudfoundry/gorouter/blob/8cd1dafe08232ed38614ad00002f3d5bc3d6633b/proxy/round_tripper/proxy_round_tripper_test.go#L407-L472

Reply:
But what you don't test is if all backends fail and cause a retry. In that case it would loop indefinitely, no?

Reply:
In that case it will retry up to the MaxAttempts which is 3 by default. This is the test for this case - https://github.com/cloudfoundry/gorouter/blob/8cd1dafe08232ed38614ad00002f3d5bc3d6633b/proxy/round_tripper/proxy_round_tripper_test.go#L388-L400

Reply:
Sorry, didn't see that 🙈

My main concern is that this is still somewhat of a regression. We have set that limit to a very high value relying on the gorouter not to try any endpoint twice. If you run into overload situations gorouter would keep hammering the few instances which are already overloaded because of the high value. On the other hand, reducing the limit would cause it not to try most of the available endpoints.

Reply:
How is it a regression if it retries less than it did before?

Reply:
@maxmoehl are you saying that you have overwritten the default in your environment to be something very large? https://github.com/cloudfoundry/routing-release/blob/develop/jobs/gorouter/spec#L221

Reply:
@maxmoehl Sorry I was confused, you are right there is an issue when we keep updating the pool while the retries are happening. And thanks to @tcdowney for good questions that helped me to understand the problem.

I added [some test cases](https://github.com/cloudfoundry/gorouter/blob/retry-new-endpoints/proxy/round_tripper/proxy_round_tripper_test.go#L415-L523) which I think covers what you are pointing to.

I can't use your suggested solution [here](https://github.com/cloudfoundry/gorouter/pull/470#discussion_r2020684257) because it won't cover the case for when new instance is added and old is removed (the numEndpoints will be the same) and this is the case we are trying to solve when instance is being updated.

I added the [following logic](https://github.com/cloudfoundry/gorouter/blob/retry-new-endpoints/proxy/round_tripper/proxy_round_tripper.go#L155-L166): if the number of attempts will exceed the total number of endpoints we will check that the new selected endpoint is the one we have not tried yet, in that case we will try that new endpoint.

Reply:
@maxmoehl great! would you want to approve this PR?